vizro-core/src/vizro/_vizro.py

```diff
@@ -165,22 +172,28 @@ def _pre_build():
         # Note that a pre_build method can itself add a model (e.g. an Action) to the model manager, and so we need to
         # iterate through set(model_manager) rather than model_manager itself or we loop through something that
         # changes size.
+        # TODO[MS]: Check that this works in the case where we dynamically add to the tree during pre-build.
         # Any models that are created during the pre-build process *will not* themselves have pre_build run on them.
         # In future may add a second pre_build loop after the first one.
-
-        for filter in cast(Iterable[Filter], model_manager._get_models(Filter)):
-            # Run pre_build on all filters first, then on all other models. This handles dependency between Filter
-            # and Page pre_build and ensures that filters are pre-built before the Page objects that use them.
-            # This is important because the Page pre_build method checks whether filters are dynamic or not, which is
-            # defined in the filter's pre_build method. Also, the calculation of the data_frame Parameter targets
-            # depends on the filter targets, so they should be pre-built after the filters as well.
-            # It's also essential for filters to be pre-built before Container.pre_build runs or otherwise
-            # control.selector won't be set.
-            filter.pre_build()
-        for model_id in set(model_manager):
-            model = model_manager[model_id]
-            if hasattr(model, "pre_build") and not isinstance(model, Filter):
-                model.pre_build()
+        # MS: I believe that this currently holds only for things that are created below in the tree.
```

Comment:
Hmm, it would be good to check this with some basic toy example (just nutree outside Vizro). What happens if you add a node to the tree during iteration - does the new node get visited or not? Does it depend on the traversal order?